File: apps/web/types/telecalling.ts
/** Mirrors apps/api/app/schemas/telecalling.py. */

export type TelecallingStatus =
  | "not_called"
  | "call_attempted"
  | "connected"
  | "not_received"
  | "busy"
  | "switched_off"
  | "invalid_number"
  | "call_back_requested"
  | "interested"
  | "not_interested"
  | "follow_up_required"
  | "confirmed"
  | "cancelled"

export const CALL_OUTCOME_OPTIONS: { label: string; value: TelecallingStatus }[] = [
  { label: "Call Attempted", value: "call_attempted" },
  { label: "Connected", value: "connected" },
  { label: "Not Received", value: "not_received" },
  { label: "Busy", value: "busy" },
  { label: "Switched Off", value: "switched_off" },
  { label: "Invalid Number", value: "invalid_number" },
  { label: "Call Back Requested", value: "call_back_requested" },
  { label: "Interested", value: "interested" },
  { label: "Not Interested", value: "not_interested" },
  { label: "Follow-up Required", value: "follow_up_required" },
  { label: "Confirmed", value: "confirmed" },
  { label: "Cancelled", value: "cancelled" },
]

/** Mirrors app.models.enums.LeadCategory. */
export type LeadCategory =
  "abandoned_checkout" | "cod_unfulfilled" | "cod_fulfilled" | "prepaid"

/** Mirrors app.models.enums.LeadPriority. */
export type LeadPriority = "high" | "medium" | "low"

export const LEAD_CATEGORY_OPTIONS: { label: string; value: LeadCategory }[] = [
  { label: "COD Unfulfilled", value: "cod_unfulfilled" },
  { label: "COD Fulfilled", value: "cod_fulfilled" },
  { label: "Prepaid", value: "prepaid" },
]

export const LEAD_CATEGORY_LABELS: Record<LeadCategory, string> = {
  abandoned_checkout: "Abandoned Checkout",
  cod_unfulfilled: "COD Unfulfilled",
  cod_fulfilled: "COD Fulfilled",
  prepaid: "Prepaid",
}

export const LEAD_PRIORITY_LABELS: Record<LeadPriority, string> = {
  high: "High",
  medium: "Medium",
  low: "Low",
}

export interface AssignedOrder {
  order_id: string
  order_number: string
  customer_name: string | null
  customer_phone: string | null
  item_summary: string | null
  total_amount: string
  payment_type: string
  payment_status: string
  fulfillment_status: string
  order_datetime: string
  shipping_address: {
    line1?: string
    line2?: string
    city?: string
    state?: string
    pin_code?: string
    country?: string
    contact_name?: string
    contact_phone?: string
  } | null
  assignment_id: string | null
  assigned_to: string | null
  assigned_to_name: string | null
  call_status: TelecallingStatus | null
  attempt_count: number
  last_attempt_at: string | null
  next_follow_up_at: string | null
  lead_category: LeadCategory | null
  priority: LeadPriority | null
}

export interface AssignedCheckout {
  checkout_id: string
  customer_name: string | null
  customer_phone: string | null
  customer_email: string | null
  item_summary: string | null
  total_amount: string
  checkout_url: string | null
  checkout_created_at: string | null
  is_recovered: boolean
  assignment_id: string | null
  assigned_to: string | null
  assigned_to_name: string | null
  call_status: TelecallingStatus | null
  attempt_count: number
  last_attempt_at: string | null
  next_follow_up_at: string | null
  lead_category: "abandoned_checkout"
  priority: LeadPriority | null
}

export interface CheckoutAssignment {
  id: string
  checkout_id: string
  assigned_to: string
  assigned_by: string | null
  assigned_at: string
  team_leader_id: string | null
  assignment_status: "active" | "inactive"
  reassigned_from: string | null
  reassigned_to: string | null
  reassigned_at: string | null
  reassignment_reason: string | null
  current_status: TelecallingStatus
  attempt_count: number
  last_attempt_at: string | null
  next_follow_up_at: string | null
}

export interface CheckoutCallAttempt {
  id: string
  checkout_id: string
  telecaller_id: string | null
  attempt_number: number
  attempted_at: string
  outcome: TelecallingStatus
  notes: string | null
  next_follow_up_at: string | null
  created_at: string
}

export interface AssignCheckoutsInput {
  checkout_ids: string[]
  mode: "manual" | "equal"
  telecaller_id?: string
  telecaller_ids?: string[]
}

export interface ReassignCheckoutInput {
  checkout_id: string
  new_telecaller_id: string
  reason: string
}

export interface CallAttempt {
  id: string
  order_id: string
  telecaller_id: string | null
  attempt_number: number
  attempted_at: string
  outcome: TelecallingStatus
  notes: string | null
  next_follow_up_at: string | null
  created_at: string
}

export interface CallHistoryEntry extends CallAttempt {
  order_number: string
}

export interface OrderAssignment {
  id: string
  order_id: string
  assigned_to: string
  assigned_by: string | null
  assigned_at: string
  team_leader_id: string | null
  assignment_status: "active" | "inactive"
  reassigned_from: string | null
  reassigned_to: string | null
  reassigned_at: string | null
  reassignment_reason: string | null
  current_status: TelecallingStatus
  attempt_count: number
  last_attempt_at: string | null
  next_follow_up_at: string | null
}

/** One selectable entry for a "Select Telecaller" assignment dropdown —
 * every active TELECALLER-role user in scope, regardless of whether
 * they've ever had a lead assigned yet. Distinct from
 * `TelecallerPerformance` below (assignment-activity counts only, so a
 * brand-new Telecaller never appears there).
 */
export interface TelecallerOption {
  id: string
  name: string
  email: string
}

export interface TelecallerPerformance {
  telecaller_id: string
  telecaller_name: string
  assigned: number
  called: number
  pending: number
  connected: number
  interested: number
  follow_ups: number
  confirmed: number
  not_interested: number
  conversion_rate: number
}

export interface TelecallingSummary {
  total_leads: number
  unassigned_leads: number
  assigned: number
  pending: number
  called: number
  connected: number
  follow_ups_today: number
  confirmed: number
  not_interested: number
  abandoned_checkouts: number
  cod_unfulfilled: number
  cod_fulfilled: number
  prepaid: number
}

export interface AssignOrdersInput {
  order_ids: string[]
  mode: "manual" | "equal"
  telecaller_id?: string
  telecaller_ids?: string[]
}

export interface ReassignOrderInput {
  order_id: string
  new_telecaller_id: string
  reason: string
}

export interface LogCallInput {
  outcome: TelecallingStatus
  notes?: string
  next_follow_up_at?: string
}

File: apps/api/app/schemas/telecalling.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import (
    AddressValidationStatus,
    FulfillmentStatus,
    LeadCategory,
    LeadPriority,
    OrderStatus,
    PaymentStatus,
    PaymentType,
    TelecallingStatus,
)


class OrderAddressUpdateRequest(BaseModel):
    """Editable shipping-address fields — exactly `Order.shipping_
    address`'s existing dict shape (see `app.integrations.shopify.
    normalizer.normalize_address`), never a new/duplicate address
    concept. Field widths match that same normalizer's `max_len` values.
    """

    contact_name: str | None = Field(default=None, max_length=255)
    contact_phone: str | None = Field(default=None, max_length=32)
    line1: str = Field(min_length=1, max_length=255)
    line2: str | None = Field(default=None, max_length=255)
    city: str = Field(min_length=1, max_length=120)
    state: str | None = Field(default=None, max_length=120)
    pin_code: str = Field(min_length=1, max_length=16)
    country: str = Field(min_length=1, max_length=120, default="India")


class AssignOrdersRequest(BaseModel):
    order_ids: list[uuid.UUID] = Field(min_length=1)
    mode: Literal["manual", "equal"]
    telecaller_id: uuid.UUID | None = None
    telecaller_ids: list[uuid.UUID] | None = None

    @model_validator(mode="after")
    def _check_mode_fields(self) -> AssignOrdersRequest:
        if self.mode == "manual" and not self.telecaller_id:
            raise ValueError("telecaller_id is required for manual assignment.")
        if self.mode == "equal" and not self.telecaller_ids:
            raise ValueError("telecaller_ids is required for equal-distribution assignment.")
        return self


class ReassignOrderRequest(BaseModel):
    order_id: uuid.UUID
    new_telecaller_id: uuid.UUID
    reason: str = Field(min_length=1, max_length=1000)


class AssignCheckoutsRequest(BaseModel):
    checkout_ids: list[uuid.UUID] = Field(min_length=1)
    mode: Literal["manual", "equal"]
    telecaller_id: uuid.UUID | None = None
    telecaller_ids: list[uuid.UUID] | None = None

    @model_validator(mode="after")
    def _check_mode_fields(self) -> AssignCheckoutsRequest:
        if self.mode == "manual" and not self.telecaller_id:
            raise ValueError("telecaller_id is required for manual assignment.")
        if self.mode == "equal" and not self.telecaller_ids:
            raise ValueError("telecaller_ids is required for equal-distribution assignment.")
        return self


class ReassignCheckoutRequest(BaseModel):
    checkout_id: uuid.UUID
    new_telecaller_id: uuid.UUID
    reason: str = Field(min_length=1, max_length=1000)


class LogCallRequest(BaseModel):
    outcome: TelecallingStatus
    notes: str | None = Field(default=None, max_length=4000)
    next_follow_up_at: datetime | None = None

    @model_validator(mode="after")
    def _reject_not_called(self) -> LogCallRequest:
        if self.outcome == TelecallingStatus.NOT_CALLED:
            raise ValueError("NOT_CALLED is not a loggable call outcome.")
        return self


class ScheduleFollowUpRequest(BaseModel):
    next_follow_up_at: datetime


class BulkConfirmOrdersRequest(BaseModel):
    order_ids: list[uuid.UUID] = Field(min_length=1, max_length=200)


class BulkConfirmOrderResult(BaseModel):
    order_id: uuid.UUID
    success: bool
    message: str | None = None


class BulkConfirmOrdersResponse(BaseModel):
    confirmed_count: int
    failed_count: int
    results: list[BulkConfirmOrderResult]


class CallAttemptResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order_id: uuid.UUID
    telecaller_id: uuid.UUID | None
    attempt_number: int
    attempted_at: datetime
    outcome: TelecallingStatus
    notes: str | None
    next_follow_up_at: datetime | None
    created_at: datetime
    is_edited: bool = False


class OrderAssignmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order_id: uuid.UUID
    assigned_to: uuid.UUID
    assigned_by: uuid.UUID | None
    assigned_at: datetime
    team_leader_id: uuid.UUID | None
    assignment_status: str
    reassigned_from: uuid.UUID | None
    reassigned_to: uuid.UUID | None
    reassigned_at: datetime | None
    reassignment_reason: str | None
    current_status: TelecallingStatus
    attempt_count: int
    last_attempt_at: datetime | None
    next_follow_up_at: datetime | None
    created_at: datetime
    updated_at: datetime


class AssignedOrderItemResponse(BaseModel):
    """One order line item, enough to show "[image] product name /
    variant title, SKU: ..." on the Telecaller order detail page — see
    `app.api.v1.endpoints.team.to_assigned_order_response`. `sku`/
    `product_name`/quantity/price fields are `OrderItem` columns
    (verbatim, never invented); `variant_title`/`image_url` come from the
    related `ProductVariant`/`Product` (via `OrderItem.product_variant`)
    and are `None` whenever that relationship never resolved (unsynced
    SKU, manual order, or — for `image_url` — a product with no Shopify
    image) rather than a data gap.
    """

    id: uuid.UUID
    sku: str
    product_name: str
    variant_title: str | None = None
    image_url: str | None = None
    quantity: int
    unit_price: Decimal
    total_amount: Decimal


class AssignedOrderResponse(BaseModel):
    """One row of a Team Leader's/Telecaller's order list — flattens the
    order + its active assignment into the exact columns the spec's
    tables ask for, the same "denormalize once, no N+1" convention as
    `OrderListResponse` (`app/schemas/order.py`).
    """

    order_id: uuid.UUID
    order_number: str
    customer_name: str | None
    customer_phone: str | None
    item_summary: str | None
    total_amount: Decimal
    payment_type: PaymentType
    payment_status: PaymentStatus
    # `Order.status` — the order-confirmation/pack-ship workflow. Never
    # the same fact as `call_status` below (`TelecallingStatus`, a call
    # outcome) — the frontend must not conflate the two.
    status: OrderStatus
    fulfillment_status: FulfillmentStatus
    confirmed_at: datetime | None = None
    confirmed_by_telecaller_id: uuid.UUID | None = None
    order_datetime: datetime
    shipping_address: dict | None = None
    # Outbound OMS -> Shopify sync state for `shipping_address` above
    # (`ShopifyFulfillmentService.sync_shipping_address`) -- lets the
    # frontend tell "saved and synced to Shopify" apart from "saved in
    # OMS, Shopify sync failed" after an address edit, rather than
    # reporting blanket success either way.
    shipping_address_sync_status: str | None = None
    shipping_address_sync_error: str | None = None
    # Same already-loaded `Order` row as everything else on this
    # response -- zero extra queries. `None` means "not yet validated",
    # shown by the frontend as "Validation pending", never guessed.
    shipping_address_validation_status: AddressValidationStatus | None = None
    shipping_address_validation_score: int | None = None
    # Line items for the Product section (image + name + variant + SKU) —
    # see `AssignedOrderItemResponse`.
    items: list[AssignedOrderItemResponse] = []
    # Null across this whole block means "not yet assigned to anyone" —
    # only possible in the Team Leader's unfulfilled-orders *pool* view
    # (`GET /team/orders/unfulfilled` with no `telecaller_id` filter),
    # which deliberately surfaces the unassigned backlog alongside
    # already-assigned-within-team orders so there's something to select
    # and assign in the first place. Every other list (a telecaller's own
    # orders, one telecaller's workload) only ever returns assigned rows.
    assignment_id: uuid.UUID | None = None
    assigned_to: uuid.UUID | None = None
    assigned_to_name: str | None = None
    call_status: TelecallingStatus | None = None
    attempt_count: int = 0
    last_attempt_at: datetime | None = None
    next_follow_up_at: datetime | None = None
    # `None` only for `PaymentType.OTHER` — not one of the spec's defined
    # categories (see `app.services.lead_classification.classify_order`).
    lead_category: LeadCategory | None = None
    priority: LeadPriority | None = None


class AssignedCheckoutResponse(BaseModel):
    """`AssignedOrderResponse`'s counterpart for an abandoned-checkout
    lead — same "denormalize once" shape, flattening an `AbandonedCheckout`
    + its (possibly absent) active `CheckoutAssignment` into one row.
    `lead_category` is always `ABANDONED_CHECKOUT` here (included anyway
    so both lead types share one column set in the frontend's unified
    lead table).
    """

    checkout_id: uuid.UUID
    customer_name: str | None
    customer_phone: str | None
    customer_email: str | None
    item_summary: str | None = None
    total_amount: Decimal
    checkout_url: str | None
    checkout_created_at: datetime | None
    is_recovered: bool
    assignment_id: uuid.UUID | None = None
    assigned_to: uuid.UUID | None = None
    assigned_to_name: str | None = None
    call_status: TelecallingStatus | None = None
    attempt_count: int = 0
    last_attempt_at: datetime | None = None
    next_follow_up_at: datetime | None = None
    lead_category: LeadCategory = LeadCategory.ABANDONED_CHECKOUT
    priority: LeadPriority | None = None


class CheckoutAssignmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    checkout_id: uuid.UUID
    assigned_to: uuid.UUID
    assigned_by: uuid.UUID | None
    assigned_at: datetime
    team_leader_id: uuid.UUID | None
    assignment_status: str
    reassigned_from: uuid.UUID | None
    reassigned_to: uuid.UUID | None
    reassigned_at: datetime | None
    reassignment_reason: str | None
    current_status: TelecallingStatus
    attempt_count: int
    last_attempt_at: datetime | None
    next_follow_up_at: datetime | None
    created_at: datetime
    updated_at: datetime


class CheckoutCallAttemptResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    checkout_id: uuid.UUID
    telecaller_id: uuid.UUID | None
    attempt_number: int
    attempted_at: datetime
    outcome: TelecallingStatus
    notes: str | None
    next_follow_up_at: datetime | None
    created_at: datetime


class CallHistoryEntryResponse(CallAttemptResponse):
    """One row of the Telecaller's "Call History" page — a `CallAttempt`
    plus the order number, since that page spans every order the
    telecaller has ever called (unlike the per-order call history on the
    order-detail page, which already has the order in context).
    """

    order_number: str


class TelecallerOptionResponse(BaseModel):
    """One selectable entry for a "Select Telecaller" assignment dropdown
    — deliberately just the User fields an assignment UI needs, not a
    second Telecaller data source: sourced straight from `User`/`Role`
    (Administration -> Users), independent of whether that telecaller
    has ever had a lead assigned yet.
    """

    id: uuid.UUID
    name: str
    email: str


class TelecallerPerformanceResponse(BaseModel):
    telecaller_id: uuid.UUID
    telecaller_name: str
    assigned: int
    called: int
    pending: int = 0
    connected: int
    interested: int = 0
    follow_ups: int
    # Call-outcome confirmed (`TelecallingStatus.CONFIRMED`, from a logged
    # call attempt) — NOT the same fact as `orders_confirmed` below
    # (`Order.status == CONFIRMED`, set by the telecaller Confirm action).
    # Deliberately separate fields, never merged, per the "don't mix
    # TelecallingStatus/OrderStatus/ShipmentStatus" rule.
    confirmed: int
    not_interested: int
    # Percentage (0-100, one decimal) of assigned leads marked CONFIRMED —
    # this codebase's existing enum has no separate "Converted" status
    # (spec's suggested outcome list, reused as-is per the "don't create
    # duplicate outcome systems" rule); CONFIRMED is treated as the
    # converted-lead signal.
    conversion_rate: float = 0.0
    # Live snapshot counts from `Order.confirmed_by_telecaller_id` /
    # `Shipment.current_status` — see
    # `OrderRepository.telecaller_confirmation_counts`'s docstring.
    orders_confirmed: int = 0
    shipped: int = 0
    delivered: int = 0


class TelecallerDetailSummaryResponse(BaseModel):
    """Summary cards for one telecaller's individual performance page —
    `TelecallerPerformanceResponse`'s fields plus `cancelled` and
    `fulfilled` (a live count, not a historical one — see
    `TelecallingService.telecaller_detail_summary`'s docstring for why no
    day-bucketed fulfillment figure is offered).
    """

    telecaller_id: uuid.UUID
    telecaller_name: str
    assigned: int
    called: int
    pending: int = 0
    connected: int
    interested: int = 0
    follow_ups: int
    confirmed: int
    not_interested: int
    cancelled: int = 0
    fulfilled: int = 0
    # Real total call-attempt count (an order called 3 times counts as 3
    # here, not 1) — distinct from `called`, which is a per-order "has
    # this order been called at least once" count. See
    # `OrderAssignmentRepository.total_attempt_count`'s docstring.
    total_attempts: int = 0
    # `total_attempts / orders_with_attempts` -- the dashboard's "Average
    # Call Attempts" tile. 0.0 when nothing has been called yet. See
    # `OrderAssignmentRepository.attempt_averaging_stats`'s docstring.
    average_call_attempts: float = 0.0
    conversion_rate: float = 0.0
    # Live snapshot counts from `Order.confirmed_by_telecaller_id` /
    # `Shipment.current_status` — see
    # `OrderRepository.telecaller_confirmation_counts`'s docstring for why
    # these are current-state counts, not a day-bucketed history.
    orders_confirmed: int = 0
    shipped: int = 0
    delivered: int = 0
    ndr: int = 0
    rto: int = 0


class TelecallerDailyPerformancePoint(BaseModel):
    date: str
    attempts: int
    connected: int
    confirmed: int
    not_interested: int
    cancelled: int
    follow_ups: int


class TelecallingSummaryResponse(BaseModel):
    total_leads: int = 0
    unassigned_leads: int = 0
    assigned: int
    pending: int
    called: int
    connected: int
    follow_ups_today: int
    confirmed: int
    not_interested: int = 0
    abandoned_checkouts: int = 0
    cod_unfulfilled: int = 0
    cod_fulfilled: int = 0
    prepaid: int = 0
